src/librustc_mir/interpret/terminator.rs

```diff
@@ -284,6 +284,18 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             ty::InstanceDef::DropGlue(..) |
             ty::InstanceDef::CloneShim(..) |
             ty::InstanceDef::Item(_) => {
+                // If this function is a `const fn` then as an optimization we can query this
+                // evaluation immediately.
+                //
+                // For the moment we only do this for functions which take no arguments
+                // (or all arguments are ZSTs) so that we don't memoize too much.
+                if self.tcx.is_const_fn_raw(instance.def.def_id()) &&
```

Comment:
This means we do that even in the Miri interpreter, when such functions are run at "run-time", right? I think that's wrong: the function, even without any arguments, can access global state and cause aliasing violations on each call. So I think this PR as-is will regress Miri's ability to find UB.

Reply:
I propose to instead move memoization into const_eval's `find_fn` implementation.

Reply:
Is there a situation where a const fn can access global state which itself is not constant? This seems like the only situation to me where aliasing may be an issue? (But I think you know much more about Miri than me! 😄)

Reply:
> I propose to instead move memoization into const_eval's find_fn implementation.

How would such an implementation work? Memoize the evaluation of the body and then return that from find_fn? Or memoize something else entirely?

Reply:
Well we might have to expand what `find_fn` can return, and the name might have to change reflecting that.

> Is there a situation where a const fn can access global state which itself is not constant? 

Right now, probably not, but as `const fn` get more powerful and allow `unsafe` code, that seems very possible to me. https://github.com/rust-lang/rust/pull/66302 is a step in that direction. At that point CTFE relies on *dynamic checks* (at CTFE evaluation time), which Miri of course doesn't do as it runs run-time code.

Reply:
> Right now, probably not, but as const fn get more powerful and allow unsafe code, that seems very possible to me.

Yikes - I wasn't aware we'll one day have unsafe code in const fn!

> Well we might have to expand what find_fn can return, and the name might have to change reflecting that.

Ok. I'm going to be unavailable for the next week but will follow up on rust-lang/miri#1081 on my return.

Reply:
> Right now, probably not, but as const fn get more powerful and allow unsafe code, that seems very possible to me

If we ever allowed (without it being UB) you to mutate or read mutable state in const fn, we couldn't ever do a MIR optimization that evals a const fn call where all args are known at compile-time. Also it would allow constants to read from global mutable state in its original state.

There's no need for dynamic checks if we do post monomorphization checks for this situation.

Reply:
> If we ever allowed (without it being UB) you to mutate or read mutable state in const fn, we couldn't ever do a MIR optimization that evals a const fn call where all args are known at compile-time. Also it would allow constants to read from global mutable state in its original state.

We certainly shouldn't allow that during CTFE, but that's not what I talked about. I am talking about Miri-the-standalone-tool.

> There's no need for dynamic checks if we do post monomorphization checks for this situation.

Now I am very confused, the dynamic checks are post-monomorphization? Both are the same thing to me in this context. Also I thought we all agreed that we should check everything dynamically so that even "Miri unleashed" is *sound*?

I thought we had consensus that the dynamic check [described all the way at the bottom here](https://github.com/rust-lang/const-eval/blob/master/const.md#reading-statics) should definitely be added. (Not that that is very relevant to this PR, but you brought it up.)

Reply:
Right, I don't remember what I meant by that.